pkg/xds/generator/zoneproxy/generator.go

```diff
@@ -0,0 +1,151 @@
+		})
+	}
+
+	return resources, nil
+}
+
+// AddFilterChains adds filter chains to a listener. Generated listener assumes that
+// mTLS is on. Using TLSInspector we sniff SNI value. SNI value has service name
+// and tag values specified with the following format:
+// "backend{cluster=2,version=1}". We take all possible destinations from
+// TrafficRoutes + MeshHTTPRoutes + GatewayRoutes and generate
+// FilterChainsMatcher for each unique destination. This approach has
+// a limitation: additional tags on outbound in Universal mode won't work across
+// different zones. Traffic is NOT decrypted here, therefore we don't need
+// certificates and mTLS settings.
+func AddFilterChains(
+	availableServices []*mesh_proto.ZoneIngress_AvailableService,
+	apiVersion core_xds.APIVersion,
+	listenerBuilder *envoy_listeners.ListenerBuilder,
+	destinationsPerService map[string][]envoy_tags.Tags,
+	endpointMap core_xds.EndpointMap,
+) {
+	sniUsed := map[string]struct{}{}
+	for _, service := range availableServices {
+		serviceName := service.Tags[mesh_proto.ServiceTag]
+		destinations := destinationsPerService[serviceName]
+		destinations = append(destinations, destinationsPerService[mesh_proto.MatchAllTag]...)
+		clusterName := envoy_names.GetMeshClusterName(service.Mesh, serviceName)
+		serviceEndpoints := endpointMap[serviceName]
+
+		for _, destination := range destinations {
+			sni := tls.SNIFromTags(destination.
+				WithTags(mesh_proto.ServiceTag, serviceName).
+				WithTags("mesh", service.Mesh),
+			)
+			if _, ok := sniUsed[sni]; ok {
+				continue
+			}
+			sniUsed[sni] = struct{}{}
+
+			// relevantTags is a set of tags for which it actually makes sense to do LB split on.
+			// If the endpoint list is the same with or without the tag, we should just not do the split.
+			// However, we should preserve full SNI, because the client expects Zone Proxy to support it.
+			// This solves the problem that Envoy deduplicate endpoints of the same address and different metadata.
+			relevantTags := envoy_tags.Tags{}
+			for key, value := range destination {
+				matchedTargets := map[string]struct{}{}
+				allTargets := map[string]struct{}{}
+				for _, endpoint := range serviceEndpoints {
+					address := fmt.Sprintf("%s:%d", endpoint.Target, endpoint.Port)
+					if endpoint.Tags[key] == value || value == mesh_proto.MatchAllTag {
+						matchedTargets[address] = struct{}{}
+					}
+					allTargets[address] = struct{}{}
+				}
+				if len(matchedTargets) < len(allTargets) {
+					relevantTags[key] = value
+				}
+			}
+
+			cluster := envoy_common.NewCluster(
+				envoy_common.WithName(clusterName),
+				envoy_common.WithTags(relevantTags),
+			)
+
+			filterChain := envoy_listeners.FilterChain(
+				envoy_listeners.NewFilterChainBuilder(apiVersion, envoy_common.AnonymousResource).Configure(
+					envoy_listeners.MatchTransportProtocol("tls"),
+					envoy_listeners.MatchServerNames(sni),
+					envoy_listeners.TcpProxyDeprecatedWithMetadata(
```

Comment:
This is deprecated? Why are we still using it? Looking at the code it doesn't look like there's much info on the deprecation. @michaelbeaumont can you explain?

Reply:
It's because I [introduced Splits to unify MeshHTTP- and MeshTCPRoute](https://github.com/kumahq/kuma/pull/6943), but to rewrite everything would have been a much larger change. To be honest I forgot about including `// deprecated:` at the time.

Reply:
I've got import cycles. This requires moving some files around. I'd like to avoid it for this PR.